internal/pkg/hub/job.go

```diff
@@ -140,48 +149,115 @@ func (j *Job) Run(h *Hub, done chan bool) error {
+					log.Tracef("[%s/%d] Adding syscall '%s' to kernel-filter\n", j.JobID(), pns, sc.Name)
 
 					// This will create a sub-filter off of the pid-namespace
 					// which matches this subset of syscalls..
 					h.PushJob(j, pns, sc.Nr)
 
+					sampleRate := spec.SampleRate
+
+					if sampleRate > 0 {
+						// find the job with the lowest current sampleRate, and
+						// if it is lower than this job's sampleRate, replace it
+						// with this one. The other job's will emulate via
+						// sub-sampling the true samplerate.
+						if lowestJob := h.findLowestSampleJob(sc.Nr, pns); lowestJob != nil && lowestJob.Spec.SampleRate < sampleRate {
+							log.Tracef("[%s/%d] swapping sample-rate for currently running rule to %d\n", j.JobID(), pns, sampleRate)
+							if err := h.filter.RemoveSyscall(sc.Nr, pns); err != nil {
+								log.Warnf("Couldn't remove syscall %v\n", err)
+							}
+						}
+
+						if err := h.filter.AddSampledSyscall(sc.Nr, pns, uint64(sampleRate)); err != nil {
+							log.Warnf("[%s/%d] Error adding syscall kernel-filter for '%s'\n", j.JobID(), pns, sc.Name)
+							return err
+						}
+					}
+
 					// Tell the kernel that we wish to monitor this syscall for
 					// this given pid-namespace.
 					// Note: if the filter already exists, this acts as a NOP.
 					if err := h.filter.AddSyscall(sc.Nr, pns); err != nil {
-						log.Printf("[%s/%d] Error adding syscall kernel-filter for '%s'\n", j.JobID(), pns, sc.Name)
+						log.Warnf("[%s/%d] Error adding syscall kernel-filter for '%s'\n", j.JobID(), pns, sc.Name)
 						return err
 					}
 				}
 
 			case event.ContainerDelEvent:
-				// container that matched our labels is being removed from the cluster.
+				// container that matched our labels is being removed from the
+				// cluster. Knowing that this container no longer exists, we
+				// can remove the associated global kernel filters.
 				pns := ev.PidNamespace
 				j.RemoveContainer(ev.Pod, ev.Name)
 
 				for _, sc := range calls {
-					log.Printf("[%s/%d] removing syscall '%s' to kernel-filter\n", j.JobID(), pns, sc.Name)
+					log.Tracef("[%s/%d] removing syscall '%s' to kernel-filter\n", j.JobID(), pns, sc.Name)
 
 					if err := h.filter.RemoveSyscall(sc.Nr, pns); err != nil {
-						log.Printf("[%s/%d] failed to remove syscall '%s' from kernel-filter\n", j.JobID(), pns, sc.Name)
+						log.Warnf("[%s/%d] failed to remove syscall '%s' from kernel-filter\n", j.JobID(), pns, sc.Name)
 
 						// XXX[lz]: just continue on for now - but we should
 						// really think about what to do in cases like this as
 						// it might be dire.
 					}
 				}
 			}
-		case <-done:
+		case <-ctx.Done():
 			return nil
 		}
 	}
 }
 
 // WriteEvent writes event `ev` to all listeners of this `Job`
 func (j *Job) WriteEvent(h *Hub, ev *event.TraceEvent) {
-	h.ps.Publish(ev, pubsub.LinearTreeTraverser(
-		hashPath(swJobStream, j.JobID())))
+	if j.Spec.SampleRate > 0 {
```

Comment:
This function might be a little clearer as an early return in the case of no sampling - but again, personal preference.

Reply:
I honestly don't care which conditional comes first - I've always done more-complex > least-complex.